[WIP] Feature/subproviders/lattice support

alex-miller-0 opened the pull request:
## Description

Hello! This PR adds support for interfacing with the GridPlus [Lattice1 hardware wallet](https://gridplus.io/lattice). It was developed along with [GridPlus' `web3-react` fork](https://github.com/GridPlus/web3-react/tree/add-lattice-example) in an effort to allow integrations with the Lattice. The addition is a simple subprovider schema that can be used by other modules such as `web3-react`. It was designed to match those of Trezor and Ledger.

### About the Lattice

The [Lattice](https://gridplus.io/lattice) is a brand new, next-generation hardware wallet designed for ease of use. Its large screen makes reading transactions and messages much more user-friendly. Keys are held in an HSM and all logic executes in a secure compute environment, which is physically separated from the network interface and protected by a physical anti-tamper mesh. The separated network interface is used to connect to external applications like this one. 

The initial batch of Lattice devices is currently being manufactured and sent to presale buyers, at which point we will turn the web store back on for more orders.

For more info see https://gridplus.io/lattice

### Product Shots

Here are some shots of my development device! (Data shown is not related to this PR)

**Sign Transaction:**

![image](https://user-images.githubusercontent.com/7378490/93615146-a366f400-f998-11ea-9cef-dce707748542.png)

**Sign Personal Message:**

![image](https://user-images.githubusercontent.com/7378490/93615268-c98c9400-f998-11ea-9eab-4ea33f95c6e3.png)

## Testing instructions

You can't test this PR without a Lattice device, but I will provide information on how I tested it on my side. If you need to test this yourself, please reach out to justin.leroux@gridplus.io and we can work out how to get you a Lattice for testing.

I was able to test with `web3-react`'s `example` Next.js app which is included in [the repo](https://github.com/NoahZinsmeister/web3-react/tree/v6/example). All changes to the app itself are found in [this commit](https://github.com/NoahZinsmeister/web3-react/commit/ffabb34a46fbe375bdf74f60ec730becbef5d7fb) of our fork. This is a simple app where you just select your wallet and then test a `signPersonalMessage` request. Here are shots of my test:

**1. Message is sent to the device from the app:**

![image](https://user-images.githubusercontent.com/7378490/94950736-b73a4c00-04a8-11eb-93d8-e9c37b386bd3.png)

**2. Message is signed on device and returned to app, which checks it against the available public key for validity:**

![Screen Shot 2020-10-02 at 9 14 38 AM](https://user-images.githubusercontent.com/7378490/94950839-e05adc80-04a8-11eb-96ea-7203ac685b3e.png)

## Types of changes

<!--- What types of changes does your code introduce? Uncomment all the bullets that apply: -->


<!-- * Bug fix (non-breaking change which fixes an issue) -->

 * New feature (non-breaking change which adds functionality)

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed. **-- I can't think of how to provide tests here, as this module must be used with others in order to interact with the Lattice**
-   [ ] Update documentation as needed. **-- Need more clarity on this. I didn't want to make manual changes to `docs/reference.mdx` since it looks like that gets generated automatically. Not sure where else to provide docs.**
-   [x] Add new entries to the relevant CHANGELOG.jsons.

alex-miller-0 committed:
Adds Lattice subprovider
The Lattice1 is a new, next generate hardware wallet from GridPlus (https://gridplus.io/lattice)
This commit adds support for the Lattice in the form of a subprovider.
Firmware on the Lattice1 currently supports getAccounts, signTransaction, and signPersonalMessage.
signTypedData is not yet supportd, but will be in the future and thus is included in this API.

alex-miller-0 commented:
I see this build is failing. I tried to setup the circle build [per instructions](https://github.com/0xProject/0x-monorepo/blob/development/CONTRIBUTING.md#fix-submit-coverage-ci-failure), but I wasn't sure how to add that `COVERALLS_REPO_TOKEN` param. I didn't want to commit my repo token here and I have never used circle before (and do not have a paid account that can run the build on a `medium+` resource), so I'm not sure what to do. Would appreciate guidance here as well.

alex-miller-0 committed:
Fixes incorrectly named param for Lattice connector

dekz commented on packages/subproviders/CHANGELOG.json:
```diff
@@ -1,4 +1,13 @@
 [
+    {
+        "timestamp": 1601647297,
```
```suggestion
```

This is automatic on a publish, so need to remove otherwise our infra thinks it has been published :)

alex-miller-0 replied:
Haha yes this felt strange to me, but I was trying to work through the contribution guidelines.

Reverted in ~b58b3a6~ cc55c2e (I think you're just saying to remove `timestamp` but keep the rest of the entry -- I can also remove the entire entry).

(By the way, I can squash these commits if/when the PR is finalized/approved.)

dekz commented on packages/subproviders/src/subproviders/lattice.ts:
```diff
@@ -0,0 +1,126 @@
+
+    /**
+     * Sign a typed data message from the account specified in the `address` param.
+     * NOTE: This method is not implemented on the Lattice in v1, but we plan to support it in future versions.
```
Might be best to throw an error whilst it's not implemented? Otherwise the dev may have a hard time tracking down a bug.

If it does throw, some tools will default to `eth_sign` of the hash of the data.

alex-miller-0 replied:
So there is an error that gets thrown by the underlying SDK ([here](https://github.com/GridPlus/gridplus-sdk/blob/master/src/ethereum.js#L53)). That error will get sent up to the `_latticeConnectClient` ([here](https://github.com/GridPlus/eth-lattice-keyring/blob/master/index.js#L131) and [here](https://github.com/GridPlus/eth-lattice-keyring/blob/master/index.js#L152)) and will eventually make it to this module (and get thrown on [L123](https://github.com/0xProject/0x-monorepo/pull/2719/files#diff-1dad31a0058965df065352f990f8aa77R123)). We will be adding support in the next few months via the other two modules I linked (which we maintain). At that point, the `signMessage` functionality will be extensible to other signing schemas, should future protocols get ossified (since it will switch on that `protocol` value). 

Happy to remove it and come back, but I figured this would avoid having to go through two sequential PRs. But no problem either way!

alex-miller-0 replied:
@dekz went ahead and removed the comment in ea71cfc

We are actually working on an implementation now. It may be available by the time we get an integration out that uses this module.

alex-miller-0 committed:
Removes manual changelog timestamp

alex-miller-0 committed:
Removes comment note
We are currently implementing this feature and the note doesn't belong
here anyway, as it catches errors.
